### PR TITLE
[Backend] As a user, I can filter the report by keyword

### LINKED ISSUE #65: [Backend] As a user, I can filter the report by keyword
### Why?
When the user submits the keyword filter, the system will process all the user's report containing with the given keywords.

### Acceptance Criteria
- [x] Retrieve keyword filter from the dashboard
- [x] Filter keyword list following the given keyword
- [x] Update the list of keyword

### PR DESCRIPTION
**Pull request by Lahphim**
Resolve https://github.com/Lahphim/go-crawler-challenge/issues/65

## What happened 👀

When the user submits the keyword filter, the system will process all the user's report containing with the given keywords.

## Insight 📝

- [x] Retrieve keyword filter from the dashboard
- [x] Filter keyword list following the given keyword
- [x] Update the list of keyword

In BeeGo, they provide some operations with `icontains` (https://beego.me/docs/mvc/model/query.md#icontains)
For this `icontains`, it will be replace with `LIKE` operation.

## Proof Of Work 📹

Demo

### CODE REVIEW

File: controllers/dashboard_test.go [hunk 1]
```diff
@@ -163,6 +163,35 @@ var _ = Describe("DashboardController", func() {
 					Expect(len(matches)).To(BeZero())
 				})
 			})
+
+			Context("given `keyword` param is set", func() {
+				It("shows only `keyword` filter records", func() {
```
Comment (olivierobert) [inline controllers/dashboard_test.go]: To be correct syntaxically, it should be:

```suggestion
				It("shows only `keyword` filtered records", func() {
```

File: controllers/dashboard_test.go [hunk 2]
```diff
@@ -163,6 +163,35 @@ var _ = Describe("DashboardController", func() {
...
+					keywordFilter := "keyword_keyword_keyword"
```
Comment (olivierobert) [inline controllers/dashboard_test.go]: To make this test easier to understand, I recommend the following naming:

```suggestion
					keywordFilter := "expected_keyword"
```

File: controllers/dashboard_test.go [hunk 3]
```diff
@@ -163,6 +163,35 @@ var _ = Describe("DashboardController", func() {
...
+					user := FabricateUser(faker.Email(), faker.Password())
+					_ = FabricateKeyword(keywordFilter, faker.URL(), 0, user)
+					_ = FabricateKeyword(fmt.Sprintf("%v %v %v", faker.Word(), keywordFilter, faker.Word()), faker.URL(), 0, user)
```
Comment (olivierobert) [inline controllers/dashboard_test.go:175]: At first, it seems like there is no need to build a complex string for the keyword string. It seems over-complexifying 😆 But on close inspection, there are two types of keywords created: a simple keyword string and a complex keyword one. That's where comments can help by naming unused fabricated fixtures:

```suggestion
					// Exact match keyword
					_ = FabricateKeyword(keywordFilter, faker.URL(), 0, user)
					// Fuzzy match keyword
					_ = FabricateKeyword("%v %v %v", faker.Word(), keywordFilter, faker.Word()), faker.URL(), 0, user)
					// non-match keyword
					_ = FabricateKeyword("paragraph_paragraph_paragraph", faker.URL(), 0, user)
```

Think of future developers and also your future self by making code easier to understand.

File: controllers/dashboard_test.go [hunk 4]
```diff
@@ -163,6 +163,35 @@ var _ = Describe("DashboardController", func() {
...
+					_ = FabricateKeyword("paragraph_paragraph_paragraph", faker.URL(), 0, user)
```
Comment (olivierobert) [inline controllers/dashboard_test.go]: Again for explicitness, I recommend the following:

```suggestion
					_ = FabricateKeyword("nonexpected_keyword", faker.URL(), 0, user)
```

File: models/keyword_test.go [hunk 5]
```diff
@@ -99,6 +99,27 @@ var _ = Describe("Keyword", func() {
...
 
+			Context("given an operator with case insensitive contain `keyword`", func() {
+				It("returns only `keyword` filter records", func() {
+					containKeyword := "keyword_keyword"
+					var queryList = map[string]interface{}{"keyword__icontains": containKeyword}
+
+					user := FabricateUser(faker.Email(), faker.Password())
+					firstKeyword := FabricateKeyword(faker.Word(), faker.URL(), 0, user)
+					secondKeyword := FabricateKeyword(faker.Word(), faker.URL(), 0, user)
+					thirdKeyword := FabricateKeyword(fmt.Sprintf("%v %v %v", faker.Word(), containKeyword, faker.Word()), faker.URL(), 0, user)
```
Comment (olivierobert) [inline models/keyword_test.go]: Same as above. Think of using variabls which can help to make sense of all the fixtures

```suggestion
					matchedKeyword := FabricateKeyword(fmt.Sprintf("%v %v %v", faker.Word(), containKeyword, faker.Word()), faker.URL(), 0, user)
```

**Review by olivierobert — COMMENTED** [review]
The tests need to be more explicit. I spent way too much time to understand the intricacies of the tests 😵

File: controllers/dashboard.go [hunk 6]
```diff
@@ -46,9 +46,17 @@ func (c *DashboardController) actionPolicyMapping() {
 func (c *DashboardController) Index() {
 	web.ReadFromRequest(&c.Controller)
 
+	var keyword string
+	err := c.Ctx.Input.Bind(&keyword, "keyword")
```
Comment (junan) [inline controllers/dashboard.go]: It can be done also with `GetString`, ex: `keyword  := c.GetString("keyword")`.

**Review by junan — APPROVED** [review]

File: controllers/dashboard.go (hunk 6 above)
Comment (Lahphim) [reply controllers/dashboard.go]: It looks nicer 👍 https://github.com/Lahphim/go-crawler-challenge/pull/67/commits/3e8a6099f210cfac8313676534569be7616f2faa

**Review by olivierobert — APPROVED** [review]

**Pull request merged**